prov/efa/docs/efa_rdm_protocol_v4.md

```diff
@@ -0,0 +1,1408 @@
+
+## 4. Extra features and requests
+
+This chapter describes the extra features and requests of protocol v4.
+
+### 4.1 RDMA read based data transfer (RDMA read)
+
+The extra feature "RDMA read based data transfer" (RDMA read) was introduced together
+with protocol v4, when libfabric 1.10 was released. It was assigned ID 0.
+
+It is defined as an extra feature because there is a set of requirements (firmware,
+EFA kernel module and rdma-core) to be met before an endpoint can use the RDMA
+read capability, therefore an endpoint cannot assume the other party support RDMA read.
+
+The "RDMA read" extra feature corresponds to the following sub-protocols:
+long-read message, emulated long-read write, direct read.
+
+#### Long-read message sub-protocol
+
+The long-read message sub-protocol uses RDMA read to implement two-sided communication.
+
+The work flow of long-read message sub-protocol is illustrated in the following diagram:
+
+![long-read message](message_longread.png)
+
+There are two packet types involved in this protocol: LONGREAD_RTM and EOR (End Of Read).
+
+LONGREAD_RTM is sent by the sender to initiate the communication.
+
+Like all REQ packets, a LONGREAD_RTM consists of 3 parts: mandatoary header, REQ optional
+header and the application data. However, the application data part of a LONGREAD_RTM is
+special: it is not the data in the application's send buffer, but information of the
+sender buffer.
+
+In long-read message sub-protocol, sender need to construct an `read_iov`, which is
+an array of `efa_rma_iov` of application's send buffer. The `read_iov` is used
+as the appliction data in the LONGREAD_RTM packet.
+
+The binary format of a LONGREAD_RTM packet's mandatory header is listed in table 4.1
+
+Table: 4.1 the binary format of a LONGREAD_RTM packet's mandatory header
+
+| Name | Length (bytes) | type | C language type | Note |
+|-|-|-|-|-|
+| `type`           | 1 | integer | `uint8_t`  | part of base header |
+| `version`        | 1 | integer | `uint8_t`  | part of base header|
+| `flags`          | 2 | integer | `uint16_t` | part of base header |
+| `msg_id`         | 4 | integer | `uint32_t` | message ID |
+| `msg_length`     | 4 | integer | `uint64_t` | total length of the message |
+| `send_id`        | 4 | integer | `uint32_t` | ID of the receive operation  |
+| `read_iov_count` | 4 | integer | `uint32_t` | number of iov to read |
+
+Noticing the new field `read_iov_count`, which is number of `struct efa_rma_iov` in `read_iov`.
+
+To construct `read_iov`, sender need to make sure the send buffer is registered with EFA device and fill
+the registration key in `read_iov`.
+
+There are two ways to achieve that:
+
+First, if the buffer has already been registered with device, application will provide
+a memory descriptor along with the sender buffer, registration key can be extracted from the descriptor;
+
+Second, if the buffer has not been registered with EFA device, sender need to register the buffer,
```

Comment:
What happens when the receiver cannot register the recv buffer to do the read into? Is there a fallback protocol?

Reply:
There is no fallback protocol on the receive side.

What should happen if receive buffer registration failed is not defined in the protocol.

In our current implementation, if receive buffer registration failed because of out of resource, the application will clear MR cache and retry registration later. If registration is failed due to other reason, it will write an error CQ entry.